verbit/streaming_client.py

```diff
@@ -190,57 +201,143 @@ def _connect_websocket(self, media_config: MediaConfig, response_types: Response
+                should_retry = True
+
+            self._logger.warning(f'Error while connecting WebSocket: {repr(outcome_ex)}, {should_retry=}')
+
+            return should_retry
+
+        @retry(wait=wait_random_exponential(multiplier=0.5),
+               stop=stop_after_delay(self.max_connection_retry_seconds),
+               retry=retry_predicate)
+        def connect_and_retry():
+            self._ws_client.connect(ws_url, header=self._get_ws_connect_headers())
+
+        # try opening WebSocket connection
+        try:
             connect_and_retry()
 
-        except Exception:
-            self._logger.exception(f'Error connecting WebSocket!')
+        # catch and log retry errors
+        except tenacity.RetryError as retry_err:
+            statistics = connect_and_retry.retry.statistics
+            last_exception = retry_err.last_attempt.exception()
+            self._logger.error(f'Error while connecting WebSocket! Exceeded maximum retries and giving up.\n'
+                               f'Last attempt raised: {repr(last_exception)}\n'
+                               f'Retry {statistics=}')
+
+            # raise further so that exception can be handled
+            raise
+
+        # catch and log all other exceptions
+        except Exception as ex:
+            self._log_exception('Error while connecting WebSocket', ex)
             raise
 
+    def _log_exception(self, msg: str, ex: Exception):
+        self._logger.error(f'{msg}: {repr(ex)}')
+        self._logger.debug(f'{msg}, stack trace:', exc_info=True)
+
     def _default_on_media_error(self, err: Exception):
-        self._logger.exception(f'Exception on media thread!')
+        self._log_exception('Exception on media thread', err)
+
+    def _send_event(self, event: str, payload: dict = None):
+
+        # use default payload if not provided
+        payload = payload or dict()
+
+        # prepare message dict
+        msg = dict(event=event, payload=payload)
+
+        # serialize as json
+        msg_json = json.dumps(msg)
 
-    def _media_sender_worker(self, media_generator: typing.Generator):
+        # send to server
+        self._ws_client.send(msg_json)
+
+    def _media_sender_worker(self, media_generator: typing.Iterator[bytes]):
         """Thread function for emitting media from a user-given generator."""
 
         try:
 
-            self._logger.debug('Starting media stream')
+            # capture WebSocket, so that connect changes in other threads do not affect this loop
+            ws_client = self._ws_client
 
             # iterate media generator
             for chunk in media_generator:
 
                 # emit media chunk
-                self._ws_client.send_binary(chunk)
+                ws_client.send_binary(chunk)
 
-            # signal end of stream
-            self.send_event(event=self.EVENT_EOS)
             self._logger.debug(f'Finished sending media')
 
+            # signal end of stream
+            self.send_eos_event()
```

Comment:
EOS will be send to the wrong socket here, in the case of a rouge thread not dying yet, and it does happen.
Why is there a need for a separate method for `send_eos`? my refactor with the private method `_send_event()` was originally about enabling sending an event directly to a web-socket (rather to the member `_ws_client`, which is not correct behavior here)

Reply:
That is: I might be missing something : I'll make the point clearer, maybe we should add this comment in the line at the start of the media thread - the reason for Capturing the websocket  `ws_client = self._ws_client`:

The important thing - is not what will happen with the 'old web-socket', in case there's a new one (in case `self._ws_client` is already a new connection), anything regarding the old-socket - will eventually fail...

My point was about sending an `EOS` to the _new_ socket by mistake...

actually : I see another issue: consuming from that same 'media_generator' here....

This actually might solve this better... I'll try making a suggestion commit

Reply:
Ok, I see why this will not happen: or almost no harm in case it does: that case is the the `media_generator` has actually finished iteration. (no exception), however, this could start leading to hard-find situations down the road... two threads sending the EOS to the WS, etc... 

Also now that the `media_generator` is share with the `reconnect` class (and thus with new media-threads reading from it), there should be a way communicating to the thread we want is to stop (other than assuming "the websocket it holds will fail now")

Either making the media-generator a class member, and switching it to an empty-generator, or just having some flag...

Reply:
Ok, I think I understand the issue. Let's discuss f2f